storage-proofs/src/piece_inclusion_proof.rs

```diff
@@ -1,255 +1,473 @@
+        let capacity_at_pos = Self::subtree_capacity(position, tree_len);
 
-    let mut hasher = H::Function::default();
+        Self::is_pow2(capacity_at_pos) && capacity_at_pos >= length
+    }
 
-    let mut last_row = Vec::new();
+    fn height_for_length(n: usize) -> usize {
+        (n as f64).log2().ceil() as usize
+    }
 
-    for chunk in piece_data.chunks(32) {
-        last_row.push(H::Domain::try_from_bytes(chunk)?);
+    fn is_pow2(n: usize) -> bool {
+        n.count_ones() == 1
     }
 
-    for (height, ((l_hash, l_is_left), (r_hash, r_is_left))) in proof_vecs.enumerate() {
-        let mut row = Vec::new();
-        if !*l_is_left {
-            row.push(*l_hash);
-        };
-        row.extend(last_row);
-        if *r_is_left {
-            row.push(*r_hash);
+    fn subtree_capacity(pos: usize, total: usize) -> usize {
+        assert!(pos < total, "position must be less than tree capacity");
+
+        let mut capacity = 1;
+        // If tree is not 'full', then pos 0 will have subtree_capacity greater than size of tree.
+        let mut cursor = pos + next_pow2(total);
+
+        while cursor & 1 == 0 {
+            capacity *= 2;
+            cursor >>= 1;
         }
-        last_row = hash_pairs::<H>(&mut hasher, row.as_slice(), height)?;
+        capacity
     }
-    assert_eq!(last_row.len(), 1);
-    Ok(last_row[0])
 }
 
-/// For each successive pair of input hashes in row, construct a new hash according to the method used by `hasher.node`,
-/// returning a vector of the constructed hashes, in order.
-/// The result will be an error (resulting in a failed proof) if row does not contain an even number of hashes.
-fn hash_pairs<H: Hasher>(
-    hasher: &mut H::Function,
-    row: &[H::Domain],
-    height: usize,
-) -> Result<Vec<H::Domain>> {
-    let hashed: Result<Vec<_>> = row
-        .chunks(2)
-        .map(|pair| {
-            if pair.len() != 2 {
-                // If input is malformed, return an Err, which will fail the proof.
-                return Err(Error::MalformedInput);
-            }
-            hasher.reset();
-            Ok(hasher.node(pair[0], pair[1], height))
-        })
-        .collect();
+#[allow(dead_code)]
+fn create_piece_tree<H: Hasher>(
+    data: &[H::Domain],
+) -> merkle::MerkleTree<H::Domain, H::Function, VecStore<H::Domain>> {
+    let data_size = data.len();
+    // We need to compute comm_p as a merkle root over power-of-two-sized data.
+    let tree_size = next_pow2(data_size);
+
+    // If actual data is less than tree size, pad it with zeroes.
+    if data_size < tree_size {
+        let zero = H::Domain::try_from_bytes(&[0; 32]).unwrap();
+        let mut data_copy = vec![zero; tree_size];
+        data_copy[0..data_size].copy_from_slice(data);
+        merkle::MerkleTree::<H::Domain, H::Function, VecStore<H::Domain>>::new(data_copy)
+    } else {
+        merkle::MerkleTree::<H::Domain, H::Function, VecStore<H::Domain>>::new(data.to_vec())
```

Comment:
Do we not have a way to create a merkle tree, without allocating a new vec? This just needs to be iterated over afaik.